brokers/artifacts/broker.go

```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
+		if err != nil {
+			return err
+		}
+		plugin.CachedExtensions[URL] = pluginPath
 	}
+	return nil
 }
 
-// ProcessPlugin processes metas of different plugin types and passes metas of each particular type
-// to the appropriate plugin broker
-func (b *Broker) ProcessPlugin(meta model.PluginMeta) error {
-	if !utils.IsTheiaOrVscodePlugin(meta) {
-		return nil
-	}
-
-	URLs, err := getUrls(meta)
+func (b *Broker) downloadArchive(URL string, pluginID string, workDir string) (string, error) {
+	archivePath := b.ioUtils.ResolveDestPathFromURL(URL, workDir)
+	archivePath, err := b.ioUtils.Download(URL, archivePath, true)
 	if err != nil {
-		return err
+		return "", fmt.Errorf("failed to download plugin from %s: %s", URL, err)
 	}
+	return archivePath, nil
+}
 
-	workDir, err := b.ioUtils.TempDir("", "vscode-extension-broker")
-	if err != nil {
-		return err
-	}
+func (b *Broker) injectPlugin(plugin *model.CachedPlugin, archivePath string) (string, error) {
+	pluginPath := "/plugins"
 
-	archivesPaths, err := b.downloadArchives(URLs, meta, workDir)
+	if plugin.IsRemote {
+		pluginUniqueName := utils.ConvertIDToUniqueName(plugin.ID)
+		pluginPath = filepath.Join(pluginPath, "sidecars", pluginUniqueName)
+		err := b.ioUtils.MkDir(pluginPath)
+		if err != nil {
+			return "", err
+		}
+	}
+	pluginArchiveName := b.generatePluginArchiveName(plugin, archivePath)
+	pluginArchivePath := filepath.Join(pluginPath, pluginArchiveName)
+	err := b.ioUtils.CopyFile(archivePath, pluginArchivePath)
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	err = b.injectPlugin(meta, archivesPaths)
-	return err
+	return pluginArchivePath, nil
 }
 
-func (b *Broker) downloadArchives(URLs []string, meta model.PluginMeta, workDir string) ([]string, error) {
-	paths := make([]string, 0)
-	for i, URL := range URLs {
-		archivePath := b.ioUtils.ResolveDestPathFromURL(URL, workDir)
-		b.PrintDebug("Downloading VS Code extension archive '%s' for plugin '%s' to '%s'", URL, meta.ID, archivePath)
-		b.PrintInfo("Downloading VS Code extension %d/%d for plugin '%s'", i+1, len(URLs), meta.ID)
-		archivePath, err := b.ioUtils.Download(URL, archivePath, true)
-		paths = append(paths, archivePath)
-		if err != nil {
-			return nil, fmt.Errorf("failed to download plugin from %s: %s", URL, err)
-		}
-	}
-	return paths, nil
+func (b *Broker) generatePluginArchiveName(plugin *model.CachedPlugin, archivePath string) string {
+	archiveName := filepath.Base(archivePath)
+	formattedID := strings.ReplaceAll(plugin.ID, "/", ".")
+	return fmt.Sprintf("%s.%s.%s", formattedID, b.rand.String(10), archiveName)
 }
 
-func (b *Broker) injectPlugin(meta model.PluginMeta, archivesPaths []string) error {
-	for _, path := range archivesPaths {
-		pluginPath := "/plugins"
+func convertMetasToPlugins(metas []model.PluginMeta) []model.CachedPlugin {
+	plugins := make([]model.CachedPlugin, 0)
+
+	for _, meta := range metas {
+		if !utils.IsTheiaOrVscodePlugin(meta) {
+			continue
```

Comment:
In terms of Che Plugins model, even non-Theia or non-Vscode is a Che Plugin. Please consider putting here a comment why only Theia or VsCode plugins are returned here - like it's artifacts broker and only they have such...